### PR TITLE
Calculate checksum from local file if upload optimization succeeds

### PR DESCRIPTION
**Pull request by sir-sigurd**
# Description

# TODO

- [x] Unit tests
- [x] [Changelog](../tree/master/docs/CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)
- [x] test with selector_fn

### CODE REVIEW

File: api/python/quilt3/data_transfer.py [hunk 1]
```diff
@@ -530,43 +530,61 @@ def upload_part(i, start, end):
...
+        # S3ClientProvider can't currently distinguish between a user that has PUT but not LIST permissions and a
+        # user that has no permissions. If we can't find a valid client, proceed to the upload stage anyway.
+        pass
+    else:
+        dest_size = resp["ContentLength"]
+        if dest_size != size:
+            return None
+        # XXX: shouldn't we check part sizes?
+        # XXX: we could check hashes of parts, to finish faster
+        # XXX: support other checksum algorithms?
```
Comment (nl0) [inline api/python/quilt3/data_transfer.py:566]: not sure it's worth it

File: api/python/quilt3/data_transfer.py [hunk 2]
```diff
@@ -530,43 +530,61 @@ def upload_part(i, start, end):
...
+        pass
+    except S3NoValidClientError:
+        # S3ClientProvider can't currently distinguish between a user that has PUT but not LIST permissions and a
+        # user that has no permissions. If we can't find a valid client, proceed to the upload stage anyway.
+        pass
+    else:
+        dest_size = resp["ContentLength"]
+        if dest_size != size:
+            return None
+        # XXX: shouldn't we check part sizes?
```
Comment (nl0) [inline api/python/quilt3/data_transfer.py:564]: i'd say we should, or at least check the number of parts is correct

File: api/python/quilt3/data_transfer.py [hunk 3]
```diff
@@ -530,43 +530,61 @@ def upload_part(i, start, end):
...
+        # XXX: we could check hashes of parts, to finish faster
```
Comment (nl0) [inline api/python/quilt3/data_transfer.py:565]: maybe

File: docs/CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,15 @@ Entries inside each section should be ordered by type:
 
 ## Catalog, Lambdas
 !-->
+# unreleased - YYYY-MM-DD
+## Python API
+* [Fixed] If upload optimization during `push()` succeeds the checksum is calculated from local file instead of remote file ([#3968](https://github.com/quiltdata/quilt/pull/3968))
+* [Changed] Upload optimization check now tries to use S3 SHA-256 checksum and fallbacks to ETag ([#3968](https://github.com/quiltdata/quilt/pull/3968))
```
Comment (nl0) [inline docs/CHANGELOG.md:20]:
```suggestion
* [Changed] Upload optimization check now tries to use S3 SHA-256 checksum and falls back to ETag ([#3968](https://github.com/quiltdata/quilt/pull/3968))
```

**Review by nl0 — COMMENTED** [review]
looks sane, tho i'd make the checksum matching logic stricter

File: api/python/quilt3/data_transfer.py (hunk 2 above)
Comment (sir-sigurd) [reply api/python/quilt3/data_transfer.py]: now it checks number of parts

File: api/python/quilt3/data_transfer.py (hunk 3 above)
Comment (sir-sigurd) [reply api/python/quilt3/data_transfer.py]: punted

**Review by nl0 — APPROVED** [review]
lgtm

File: api/python/tests/test_data_transfer.py
[the remark this answers is not in the document]
Comment (nl0) [reply api/python/tests/test_data_transfer.py:1]: well, yeah, they do, somewhat)

**Pull request merged**